src/components/Blog/Post/Markdown/index.tsx

```diff
@@ -1,13 +1,21 @@
 import React from 'react'
+import rehypeReact from 'rehype-react'
 
 import * as styles from './styles.module.css'
 
 interface IMarkdownProps {
-  html: string
+  htmlAst: Node
 }
 
-const Markdown: React.FC<IMarkdownProps> = ({ html }) => (
-  <div className={styles.wrapper} dangerouslySetInnerHTML={{ __html: html }} />
-)
+// Rehype's typedefs don't allow for custom components, even though they work
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const renderAst = new (rehypeReact as any)({
+  createElement: React.createElement,
+  Fragment: React.Fragment
+}).Compiler
+
+const Markdown: React.FC<IMarkdownProps> = ({ htmlAst }) => {
+  return <div className={styles.wrapper}>{renderAst(htmlAst)}</div>
```

Comment:
Instead of using `html`, we use `htmlAst` allowing us to add custom components to blog like we do with the doc.

Reply:
@rogermparent @julieg18 does it mean that we now do more JS work on the client side + potentially have to bring more dependencies?

Reply:
No, I don't think so. @rogermparent, correct me if I'm wrong, but I believe HTML rendering is all done during build time, whether we use `html` or `htmlAst`.

Reply:
yep, related question also here is what is being prefetched / fetched when we hover the side panel in the docs and eventually click to get the new page - rendered HTML or some tree that still requires post-processing?

Reply:
The answer is between the two:

- Initial page load is fully compiled HTML that is then rehydrated
- Subsequent page loads are done within Gatsby's React SPA, whether we use html or rehype

So the blogs would be using JS a little more, but regardless of the html content being an HTML string or Rehype we're already in SPA world so the difference is negligible. The only extra library pulled in is `rehype-react`, which would already be pulled in when viewing a doc and is 9kb minzipped and 20kb unpacked.

I'll run some tests and inspect the browser data, between this and the prior commit without `rehype` blogs, I've run them before but the difference is negligible.

Reply:
Weirdly enough, rehype delivers less JS on average according to lighthouse.

rehype:
![image](https://user-images.githubusercontent.com/9111807/158666071-22712afc-a91b-4f8a-9e3f-d3dba6834ea1.png)

html:
![image](https://user-images.githubusercontent.com/9111807/158666153-2c8de5c4-c4a1-47d0-bafb-0b3c941f0c44.png)

Reply:
thanks Roger!

> but regardless of the html content being an HTML string or Rehype we're already in SPA world so the difference is negligible.

agreed, but I think the initial decision behind using the `dangerousHTML` call was specifically to avoid this slippery slope of bringing more and more post processing into client (e.g. that's why we were "complicating" our plugins to operate build time vs runtime as far as I remember). Ideally we should probably see CPU load, etc (does it become less "snappy" for example on large pages, etc).

> which would already be pulled in when viewing a doc

I guess initially docs engine was using dangerous html as well, right? Not sure what was the reason to migrate it to AST.

No strong opinion on this- you folks decide based on data. Just sharing the context about those decisions the way I remember them.

Reply:
`page-data.json` with rehype (there's also an html string in there, which we should remove since it's ~20kb of data we're not using)
![image](https://user-images.githubusercontent.com/9111807/158667513-c4e43d61-ee5e-4468-9602-a7b834172c8c.png)

`page-data.json` with HTML
![image](https://user-images.githubusercontent.com/9111807/158666989-a166e566-9cbf-49f8-984d-98b32ab900d3.png)

I've copied the response strings into files to compare, and the rehype page-data saves about 12kb on the same response on the same page.

Reply:
@shcheklein that's definitely a valid concern, but from what I can tell the cost/benefit analysis favors using `htmlAst` in both engines over using `html` in the blog. As far as I understand, the reason the docs switched to `htmlAst` is because the custom components used there (particularly `Abbr` tooltips since the other complex components were introduced after) would be incredibly difficult to build and maintain in a pure HTML engine.

Reply:
@rogermparent 

> would be incredibly difficult to build and maintain in a pure HTML engine.

yep, it might be. Can we find when it was done out of curiosity?